### PR TITLE
Failing scraper slack notification

### PR DESCRIPTION
**Pull request by Gathondu**
#### What does this PR do?
Sends a slack message if a scrapper fails
#### Description of Task to be completed?
Notification on Slack of failing scraper
#### How should this be manually tested?
run the following command:
`nosetests healthtools.tests.test_scrapers:TestDoctorsScraper.test_scrapper_sends_slack_notification`
#### Screenshots (if appropriate)
image

### CODE REVIEW

**Comment by Gathondu** [comment]
@DavidLemayian this PR Depends on the Indexing PR. every change there is also here but I have updated the README on this PR

**Comment by Gathondu** [comment]
@DavidLemayian this is the other branch I keep referring to. 😄

**Comment by DavidLemayian** [comment]
@Gathondu The #9 has caused conflicts with this. Make the pull and do the necessary.

**Comment by Gathondu** [comment]
@DavidLemayian i've handled the conflicts..we good to go

File: healthtools/config.py [hunk 1]
```diff
@@ -11,14 +11,7 @@
...
     "aws_secret_access_key": os.getenv("MORPH_AWS_SECRET_KEY"),
-    "region_name": os.getenv("MORPH_AWS_REGION"),
-    # Doctors document endpoint
-    "cloudsearch_doctors_endpoint": "http://doc-cfa-healthtools-ke-doctors-m34xee6byjmzcgzmovevkjpffy.eu-west-1.cloudsearch.amazonaws.com/",
-    # Clinical document endpoint
-    "cloudsearch_cos_endpoint": "http://doc-cfa-healthtools-ke-cos-nhxtw3w5goufkzram4er7sciz4.eu-west-1.cloudsearch.amazonaws.com/",
-    # Health facilities endpoint
-    "cloudsearch_health_faciities_endpoint": "https://doc-health-facilities-ke-65ftd7ksxazyatw5fiv5uyaiqi.eu-west-1.cloudsearch.amazonaws.com",
-
+    "region_name": os.getenv("MORPH_AWS_REGION")
```
Comment (DavidLemayian) [inline healthtools/config.py]: Let's have this default to `eu-west-1`.

File: healthtools/scrapers/base_scraper.py [hunk 2]
```diff
@@ -65,6 +65,11 @@ def scrape_site(self):
                 delete_batch.extend(delete_docs)
             except Exception as err:
                 skipped_pages += 1
+                requests.post(
+                    SLACK['url'],
+                    data=json.dumps({"text": "```ERROR: scrape_site() - source: {} - page: {} - {}```".format(url, page_num, err)}),
+                    headers={'Content-Type': 'application/json'}
+                    )
```
Comment (DavidLemayian) [inline healthtools/scrapers/base_scraper.py]: Due to repeating nature, this should probably be handled as a method.

File: healthtools/scrapers/base_scraper.py [hunk 3]
```diff
@@ -207,9 +230,20 @@ def delete_elasticsearch_docs(self):
...
-                print "ERROR - delete_elasticsearch_docs() - no delete file present"
+                requests.post(
+                    SLACK['url'],
+                    data=json.dumps(
+                        {"text": "```ERROR - delete_cloudsearch_docs() - no delete file present```"}),
+                    headers={'Content-Type': 'application/json'}
+                    )
+                print "ERROR - delete_cloudsearch_docs() - no delete file present"
                 return
-            print "ERROR - delete_elasticsearch_docs() - {} - {}".format(type(self).__name__, str(err))
```
Comment (DavidLemayian) [inline healthtools/scrapers/base_scraper.py]: We should still maintain error logs being printed incase Slack is not preferred / used by other people who pick up this code. Also, dev environments might prefer not to send to Slack.

File: healthtools/tests/test_scrapers.py [hunk 4]
```diff
@@ -109,3 +110,12 @@ def test_clinical_officers_scraper_archives_to_s3(self):
...
         self.health_facilities_scraper.get_token()
         self.assertIsNotNone(self.health_facilities_scraper.access_token)
+
+    def test_scrapper_sends_slack_notification(self):
+        response = requests.post(
+            SLACK['url'],
+            data=json.dumps({"text": "*This test is passing* :tada:"}),
+            headers={'Content-Type': 'application/json'}
+            )
+        self.assertEqual(response.status_code, 200)
```
Comment (DavidLemayian) [inline healthtools/tests/test_scrapers.py]: 👍 on the test add.

**Review by DavidLemayian — CHANGES_REQUESTED** [review]

**Comment by DavidLemayian** [comment]
@Gathondu Have you also checked out whether this could be useful? - https://github.com/mathiasose/slacker_log_handler

File: healthtools/scrapers/base_scraper.py (hunk 2 above)
Comment (Gathondu) [reply healthtools/scrapers/base_scraper.py]: very true..on it

File: healthtools/scrapers/base_scraper.py (hunk 3 above)
Comment (Gathondu) [reply healthtools/scrapers/base_scraper.py]: I have maintained the printing to the terminal on all errors. That should be the merge conflict bit that i had to remove

**Comment by Gathondu** [comment]
I've checked this out `https://github.com/mathiasose/slacker_log_handler` and i noticed it uses an API KEY. From slack's documentation we need to use OAuth to get the access token and this means using a web server and I was shy to use that approach since this is just a scraper and adding a web server just to get a token was an overkill hence the webhook approach. We could use the legacy token approach but that's for testing purposes and this is how it was posting on slack: 
image
and I had no control whatsoever on the handle's title

**Comment by Gathondu** [comment]
@DavidLemayian I have already made the requested changes

File: healthtools/scrapers/base_scraper.py [hunk 5]
```diff
@@ -65,6 +65,7 @@ def scrape_site(self):
                 delete_batch.extend(delete_docs)
             except Exception as err:
                 skipped_pages += 1
+                self.post_to_slack("ERROR: scrape_site() - source: {} - page: {} - {}".format(url, page_num, err))
                 print "ERROR: scrape_site() - source: {} - page: {} - {}".format(url, page_num, err)
```
Comment (DavidLemayian) [inline healthtools/scrapers/base_scraper.py]: Considering the error message is the same printed and to Slack, should we not instead use an "error method" that sends the relevant variables?

**Review by DavidLemayian — CHANGES_REQUESTED** [review]

File: healthtools/scrapers/base_scraper.py (hunk 5 above)
Comment (Gathondu) [reply healthtools/scrapers/base_scraper.py]: Makes sense

**Pull request merged**